Test/git-issues/git-issue-1604.dfy

```diff
@@ -0,0 +1,130 @@
+// RUN: %dafny /compile:3 /rprint:"%t.rprint" "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+datatype Cell = Cell(x: int)
+
+predicate isEven(x: int) {
+  x % 2 == 0
+}
+// Ghost constraint
+type GhostEvenCell = c: Cell | isEven(c.x) witness Cell(0)
+
+// Compilable constraint
+type CompilableEvenCell = c: Cell | c.x % 2 == 0 witness Cell(0)
+
+predicate method ghostEvenCellIsOneOrMore(c: GhostEvenCell)
+{
+  (if c.x % 2 == 1 then 1/0 else c.x) > 1
+}
+predicate method compiledEvenCellIsOneOrMore(c: CompilableEvenCell)
+{
+  (if c.x % 2 == 1 then 1/0 else c.x) > 1
+}
+
+function method returnsOneIfCompilableEvenCell(c: CompilableEvenCell): int
+{
+  if c.x % 2 == 1 then 1/0 else 1
+}
+function method returnsOneIfGhostEvenCell(c: GhostEvenCell): int
+{
+  if c.x % 2 == 1 then 1/0 else 1
+}
+
+
+function method {:opaque} getOriginalSet(): set<Cell> {
+  { Cell(1), Cell(2), Cell(3), Cell(4) }
+}
+
+predicate method isSetOfGhostEvenCells(s: set<GhostEvenCell>) {
+  forall c :: c in s ==> returnsOneIfGhostEvenCell(c) == 1
+}
+predicate method isSetOfCompilableEvenCells(s: set<CompilableEvenCell>) {
+  forall c :: c in s ==> returnsOneIfCompilableEvenCell(c) == 1
+}
+
+function method isEvenCell(c: Cell): (r: bool)
+{
+  c.x == 0
+}
+
+predicate method {:opaque} isMod2(c: Cell)
+  ensures c.x % 2 == 0 <==> isMod2(c)
+ {
+  c.x % 2 == 0
+}
+
+
+method Main() {
+  var x: set<Cell> := getOriginalSet();
+  var b := true;
+
+  // This line should work because c should be of type Cell as the constraint is not compilable
```

Comment:
> // This line should work because c should be of type Cell as the constraint is not compilable

I don't get this line. 

Is the type of `c` inferred to be `GhostEvenCell` based on the call to `ghostEvenCellIsOneOrMore` ?

How is `set c: GhostEvenCell | .. :: c` compilable? Shouldn't `c : GhostEvenCell` imply that we can't generate the values of `c` in a compiled context? Does the constraint `isEvenCell(c)` make it compilable, since it is compilable and happens to constrain `c: Cell` to the same subset as `c: GhostEventCell`?